### PR TITLE
[WIP] Optionally use `mmap`ing of files with spilling

### PR DESCRIPTION
**Pull request by jakirkham**
This allows the OS to handle mapping pages between file and memory. As a result if memory is getting full and some objects are not being used, the OS can free this memory up behind the scenes. Since the scratch directories being used for spilling are meant to at least be local disks (not NFS) or often solid state, IO between disk and memory can be quite fast. So reloading spilled objects is fairly reasonable.

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/spill.py [hunk 1]
```diff
@@ -228,16 +228,21 @@ class Slow(zict.Func):
     max_weight: int | Literal[False]
     weight_by_key: dict[str, SpilledSize]
     total_weight: SpilledSize
+    memmap: bool
 
-    def __init__(self, spill_directory: str, max_weight: int | Literal[False] = False):
+    def __init__(self, spill_directory: str, max_weight: int | Literal[False] = False, memmap: bool = False):
```
Comment (jakirkham) [inline distributed/spill.py:233]: Probably we want this in the config as well

**Comment by jakirkham** [comment]
Should note this is incomplete as we need the config value noted above

File: distributed/spill.py [hunk 2]
```diff
@@ -228,16 +228,26 @@ class Slow(zict.Func):
...
+        memmap: bool = False,
+    ):
+        file_kwargs = {}
+        if memmap:
+            file_kwargs["memmap"] = memmap
         super().__init__(
             partial(serialize_bytelist, on_error="raise"),
             deserialize_bytes,
-            zict.File(spill_directory),
+            zict.File(spill_directory, **file_kwargs),  # type: ignore
```
Comment (crusaderky) [inline distributed/spill.py:245]: Style nitpick: please check for `has_zict_210` instead like it already happens above

**Comment by crusaderky** [comment]
It doesn't seem to work:
```python
import dask
import numpy
from distributed.spill import Slow

dask.config.set({"distributed.comm.compression": False})
d = Slow("zictdump", memmap=True)
d["x"] = numpy.random.random(2**27)
x = d["x"]  # RSS memory blows up by 1GB
```
The memory is allocated by `deserialize_bytes`.

Assuming the above is fixed, I can see several fundamental problems:
- It cannot possibly work together with compression. There should be checks that disable compression if mmap is enabled or vice versa. Also, the compression settings should be split between spill and network, so that you can have compression on for network transfers and off for spill. However, after #5900, this would cause uncompressed data to travel over the network even if compression is enabled for network transfers - with potentially very unpleasant performance implications.

- This causes the fd to survive fh.close(). At any one given time, this means that you'll have additional open file descriptors equal to the number of tasks that are in fast but at some point in the past went through slow. This in turn could cause the process to hit its maximum number of fds. To make things worse, this is irrespective of size - a single-byte variable will cost you 1 fd.
- SpillBuffer immediately pops the key from Slow upon transition from slow to fast. This in turn deletes the underlying file on disk. On Linux, the occupied space on disk remains there as long as there are open fds to it (see above point). 
This change causes the available disk space to be reduced, without any kind of possible insight (your local directory might as well be empty!), by the size of all tasks that are in fast but at some point in the past went through slow. This will cause the `max_spill` setting to be violated.
- It's extremely inefficient if you spill the same key again, as it will be written to disk again even if it's already there. 
- The previous point will happen *a lot*, becase the decision to spill is determined by `distributed.worker.memory.threshold`, which uses the uncapped `sum(sizeof(v) for v in data.fast.values()))` - except that data in fast may or may not be using RSS memory.
- This will cause bugous readings in the GUI and in everything that uses memory metrics in decision-making (namely: the Active Memory Manager), as it will cause process memory to be substantially lower than managed (non-spilled) memory. e.g. you could now have `sum(sizeof(v) for v in data.fast.values()))` = 50GB, but only 5GB process, 0 spill, and 5GB managed_in_memory (because managed_in_memory is capped to the process memory). As a consequence you'll see the unmanaged memory nailed to zero, which is very misleading.
 
---
#### In conclusion:
I think it would be very useful to take a step back and analyse what, exactly, we are trying to achieve here, that the current explicit spill system doesn't already do, and come up with a design as a consequence.

**Comment by jakirkham** [comment]
As noted above, I'm not sure this is ready for review. That said, appreciate the feedback

**Comment by jakirkham** [comment]
Am curious why compression wouldn't work. As compressors support the Python Buffer Protocol as does `mmap`, would expect this to work, but maybe I'm missing something

**Comment by crusaderky** [comment]
Without compression:
1. mmap creates a buffer that is not actually consuming memory
2. `deserialize_bytes` (in theory) eagerly loads the first 2 frames (header + pickle data) and unpickles them. The third frame, which contains the actual numpy buffer, is created as a PickleBuffer as a view of the mmap'ed data, so it's not loaded into memory
3. your third frame will load into memory only when something actually reads the specific area of the numpy array.
4. when the Linux kernel decides to spill out an mmap'ed page, that includes the numpy bytes.

With compression:
1. idem
2. `deserialize_bytes` (in theory) passes the 3 frames to the decompressor, which eagerly loads them into memory, creates new decompressed frames, and then descopes the input mmap'ed data.
3. you now have a vanilla numpy array, not backed by mmap.

**Comment by jakirkham** [comment]
Right decompression would force it into memory.

Thought not work meant using the `mmap` would run into some other bug.
